api/envoy/config/cluster/v3/cluster.proto

```diff
@@ -507,6 +507,42 @@ message Cluster {
       google.protobuf.UInt32Value hash_balance_factor = 2 [(validate.rules).uint32 = {gte: 100}];
     }
 
+    // Configuration for :ref:`slow start mode <arch_overview_load_balancing_slow_start>`.
+    message SlowStartConfig {
+      // Endpoint warming policy defines additional condition (apart from condition that endpoint creation duration should be within slow start window)
+      // for endpoint to enter slow start mode.
+      // Refer to :ref:`endpoint warming policy types <arch_overview_load_balancing_slow_start_endpoint_warming_policy_types>`
+      // for an explanation on each policy type.
+      enum EndpointWarmingPolicy {
+        // Endpoint would enter slow start immediately.
+        NO_WAIT = 0;
+
+        // Endpoint would enter slow start upon passing first active healthcheck.
+        WAIT_FOR_FIRST_PASSING_HC = 1;
+      }
+
+      // Size of slow start window in seconds, defaults to 0.
+      // If set to value greater than 0, the newly created host remains in slow start mode starting from it's creation time
+      // for the duration of slow start window.
+      uint32 slow_start_window = 1 [(validate.rules).uint32 = {gte: 0}];
+
+      // This :ref:`endpoint warming policy type <arch_overview_load_balancing_slow_start_endpoint_warming_policy_types>` defines condition
+      // to be satisfied for host to enter slow start mode. The value defaults to
+      // :ref:`NO_WAIT<envoy_api_enum_value_config.cluster.v3.Cluster.CommonLbConfig.SlowStartConfig.EndpointWarmingPolicy.NO_WAIT>`.
+      EndpointWarmingPolicy endpoint_warming_policy = 2
+          [(validate.rules).enum = {defined_only: true}];
+
+      // This parameter linearly affects load balancing weight of an endpoint. Defaults to 1.0.
+      // The smaller the time bias is, the less traffic would be sent to endpoint that is within slow start window.
+      // The value of time bias should be greater than or equal to 0.0 and less than or equal to 1.0.
+      // If slow start window and time bias are configured, effective weight of an endpoint would be scaled with time bias:
+      // `weight = load_balancing_weight * time_bias`
```

Comment:
Can you clarify how time itself factors into this calculation or the overall calculation? It's not completely clear to me.

Reply:
Ping

Reply:
@mattklein123 would this be better?
```
If slow start window and time bias are configured, effective weight of an endpoint would be scaled with time bias during the duration of slow start window:
`weight = load_balancing_weight * time_bias`
Once duration of slow start window elapses, hosts exists slow start and its weight is no longer being scaled with time bias parameter, the regular weight is applied.
```
I may not be understanding what else  is missing in the description. It is as simple as apply time bias to weight during slow start window and not apply the bias outside of the window.

Reply:
What I'm asking for is to update the formula to account for time. Isn't time somehow related to the final weight?

Reply:
The current implementation only checks if host creation time is within time window or not and if it is it then multiplies weight with bias, eg (shortened version):
```
      auto host_weight = hostWeight(*host);
      if (host_create_duration <= slow_start_window_ ) {

        time_bias_ = time_bias_runtime_ != nullptr ? time_bias_runtime_->value() : time_bias_;

        if (time_bias_ < 0.0) {
          ENVOY_LOG(warn, "upstream: invalid time bias supplied (runtime key {}), using 1.0",
                    time_bias_runtime_->runtimeKey());
          time_bias_ = 1.0;
        }
        if (time_bias_ > 0.0) {
          host_weight *= time_bias_;
        }
      }
```

Reply:
I see. That's not typically how slow start works. Usually the weight ramps over the slow start window. I would suggest updating the formula to account for time?

Reply:
Yes, good point. I was not sure myself, as current formula would reduce amount of traffic sent to endpoint as time within slow start window is progressing. Will fix the formula.